source/Nevermore/IQueryBuilder.cs

```diff
@@ -1,40 +1,85 @@
 using System;
 using System.Collections.Generic;
-using Nevermore.Joins;
+using Nevermore.AST;
 
 namespace Nevermore
 {
-    public interface IQueryBuilder<TRecord> where TRecord : class
+    public interface IOrderedQueryBuilder<TRecord> : IQueryBuilder<TRecord> where TRecord : class
+    {
+    }
+
+    public interface ITableSourceQueryBuilder<TRecord> : IQueryBuilder<TRecord> where TRecord : class
     {
-        IQueryGenerator QueryGenerator { get; }
+        ITableSourceQueryBuilder<TRecord> View(string viewName);
+        ITableSourceQueryBuilder<TRecord> Table(string tableName);
+        ITableSourceQueryBuilder<TRecord> Alias(string tableAlias); 
+        IQueryBuilder<TRecord> Hint(string tableHint);
+        IAliasedSelectSource AsAliasedSource();
+    }
 
+    public interface ISubquerySourceBuilder<TRecord> : IQueryBuilder<TRecord> where TRecord : class
+    {
+        ISubquerySource AsSource();
+        ISubquerySourceBuilder<TRecord> Alias(string subqueryAlias);
+    }
+
+    public interface IJoinSourceQueryBuilder<TRecord> : IQueryBuilder<TRecord> where TRecord : class
+    {
+        IJoinSourceQueryBuilder<TRecord> On(string leftField, JoinOperand operand, string rightField);
+    }
+
+    public interface IQueryBuilder<TRecord> where TRecord : class
+    {
         IQueryBuilder<TRecord> Where(string whereClause);
-        IQueryBuilder<TRecord> Where(string fieldName, SqlOperand operand, object value);
-        IQueryBuilder<TRecord> Where(string fieldName, SqlOperand operand, object startValue, object endValue);
-        IQueryBuilder<TRecord> Where(string fieldName, SqlOperand operand, IEnumerable<object> values);
-        
+        IQueryBuilder<TRecord> WhereParameterised(string fieldName, UnarySqlOperand operand, Parameter parameter);
+        IQueryBuilder<TRecord> WhereParameterised(string fieldName, BinarySqlOperand operand, Parameter startValueParameter, Parameter endValueParameter);
+        IQueryBuilder<TRecord> WhereParameterised(string fieldName, ArraySqlOperand operand, IEnumerable<Parameter> parameterNames);
+
         IOrderedQueryBuilder<TRecord> OrderBy(string orderByClause);
         IOrderedQueryBuilder<TRecord> OrderByDescending(string orderByClause);
+        IQueryBuilder<TRecord> IgnoreDefaultOrderBy();
 
-        IQueryBuilder<TRecord> Join(IJoin join);
-        IQueryBuilder<TRecord> Parameter(string name, object value);
-        IQueryBuilder<TRecord> LikeParameter(string name, object value);
-        IQueryBuilder<TRecord> LikePipedParameter(string name, object value);
-        IQueryBuilder<TRecord> View(string viewName);
-        IQueryBuilder<TRecord> Table(string tableName);
-        IQueryBuilder<TRecord> Hint(string tableHint);
-        IQueryBuilder<TRecord> NoLock();
+        IQueryBuilder<TRecord> Column(string name);
```

Comment:
Since this is interface most likely encountered by the consumer, some XML doc should be added. I'm unsure at first glance what this method does.

Reply:
Done